### PR TITLE
resolv.conf fixes

### LINKED ISSUE #1920: dhcp cleanup
dhcpcd should not use `/tmp/etc` for storing state. Should be in `/run`, probably using a resolvconf style setup as per Debian/Alpine resolvconf packahes.

### PR DESCRIPTION
**Pull request by justincormack**
Do not try to change /etc/resolv.conf
    
The filesystem is supposed to be immutable, so do not try to make
a symlink; new versions of moby tool should add one anyway. But
try to make the directory a symlink points to, assuming that it
will be on a writeable filesystem.
    
fix #1920
see also #2288
updates moby to include https://github.com/moby/tool/pull/128

cat-in-a-box

### CODE REVIEW

**Comment by justincormack** [comment]
The only user facing change should be that any image that wants to have access to `resolv.conf` must add a `binds: - /etc/resolv.conf:/etc/resolv/conf`. They should have been doing that before, but previously they had a default Google DNS if they did not, so some may have been missed.

File: pkg/dhcpcd/Dockerfile [hunk 1]
```diff
@@ -16,4 +16,4 @@ WORKDIR /
 COPY --from=mirror /out/ /
 COPY /dhcpcd.conf /usr/ /
 CMD ["/sbin/dhcpcd", "--nobackground", "-f", "/dhcpcd.conf"]
-LABEL org.mobyproject.config='{"binds": ["/var:/var", "/tmp/etc:/etc"], "capabilities": ["CAP_NET_ADMIN", "CAP_NET_BIND_SERVICE", "CAP_NET_RAW"]}'
+LABEL org.mobyproject.config='{"binds": ["/run/resolvconf:/etc"], "capabilities": ["CAP_NET_ADMIN", "CAP_NET_BIND_SERVICE", "CAP_NET_RAW"]}'
```
Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: This looks a bit odd, is there nothing else in the image's `/etc/` which is interesting?

Perhaps `/run/resolvconf/resolve.conf:/etc/resolv.conf` would be less intrusive?

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: Yeah it does look "interesting" but it is somewhat hard to change. You cannot bind mount just the file as it writes a new one, so the inode changes. It also writes some other files in there (eg `ntpd.conf`) which are fairly harmless. In future we could customise the scripts to write to more useful places but mapping `etc` has been how we did it so far.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: But no, it does not use anything else in `/etc`. `dhcp` does not do DNS or anything... If there was we could also bind that over but it seems to run fine.

Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: Does the dhcp client replace a symlink or the target? If it replaces the target you could do the same symlink to `/run` trick as you do in init.

In any case, I'm ok with this for now.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: It probably replaces the target, but its really difficult to make `/etc/resolv.conf` a symlink in a container except by changing it at runtime or making them all symlinks in every image, which is also a bit weird. Going to leave for now...

Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: I was thinking that dhcp clients were a special case worth modifying in this way, but fine to leave it.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: I dont know how to tell if something is a dhcp client, other than adding a special label...

Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: I just meant to suggest adding the /run bind to the binds in the label in `pkg/dhcpcd/Dockerfile`, which you know is a client, not to do something clever elsewhere.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: You can't create a symlink with a bind as `runc` dereferences them so it still needs to have the symlink at `/etc/resolv.conf`. It is all a bit gross.

Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: I think we are talking at cross purposes, all I was suggesting was that the `pkg/dhcpcd/Dockerfile` have in it:
```
RUN ln -nfs /run/resolvconf/resolve.conf /etc/resolv.conf
LABEL org.mobyproject.config='{"binds": ["/run/resolvconf:/run/resolvconf"], ...
```
Then the dhcpc client writes through it's local `/etc/resolv.conf` into the host/shared `/run/resolvconf/resolve.conf` which the host then picks up via it's own symlink. Other "normal" packages just bind `/etc/resolv.conf:/etc/resolv.conf` as normaly and rely on the symlink deref behaviour of `runc` to get the right thing.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: you cannot have `RUN ln -nfs /run/resolvconf/resolve.conf /etc/resolv.conf` in any Dockerfile as `/etc/resolv/conf` is an external read only bind mount that you cannot override.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: (It would help if `--network-none` builds/runs did not bind mount that in)

Comment (ijc) [inline pkg/dhcpcd/Dockerfile]: > external read only bind mount that you cannot override

Damn, I always forget that :-(

In theory we could probably configure dhcpcd to write to `/etc/resolv.conf.rw` (or just to `/run` directly) but that's probably going too far down this particular rabbit hole.

Comment (justincormack) [inline pkg/dhcpcd/Dockerfile]: I think not adding the bind mounts for network stuff with network=none is the best long term solution. They are of no real use anyway, and we have network=none on build too.

**Pull request merged**